File: Data Structures & Algorithms/online-stock-span/submission-0.java
class StockSpanner {

    List<int[]> stack;

    public StockSpanner() {
        this.stack = new ArrayList<>(10);
    }

    public int next(int price) {
        if (this.stack.isEmpty()) {
            this.stack.add(new int[] { price, -1 });
            return 1;
        }

        int[] top = this.stack.getLast();
        if (top[0] > price) {
            this.stack.add(new int[] { price, this.stack.size() - 1 });
            return 1;
        } else {
            int i = this.stack.getLast()[1];
            for (;;) {
                if (i == -1) {
                    this.stack.add(new int[] { price, -1 });
                    return this.stack.size();
                }
                int[] prev = this.stack.get(i);
                if (prev[0] > price) {
                    this.stack.add(new int[] { price, i });
                    return this.stack.size() - 1 - i;
                } else {
                    i = prev[1];
                }
            }
        }
    }
}

/**
 * Your StockSpanner object will be instantiated and called as such:
 * StockSpanner obj = new StockSpanner();
 * int param_1 = obj.next(price);
 */

/**
 * Your StockSpanner object will be instantiated and called as such:
 * StockSpanner obj = new StockSpanner();
 * int param_1 = obj.next(price);
 */

File: Data Structures & Algorithms/range-sum-query-2d-immutable/submission-0.py
class NumMatrix:

    def __init__(self, matrix: List[List[int]]):
        
        rows = len(matrix)
        cols = len(matrix[0])

        new_matrix = []

        for r in range(rows + 1):
            row = [0] * (cols + 1)
            new_matrix.append(row)

        for r in range(rows):
            prefix_sum = 0
            for c in range(cols):
                prefix_sum += matrix[r][c]
                new_matrix[r + 1][c + 1] = prefix_sum + new_matrix[r][c+1]
        self.matrix = new_matrix

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        row1 += 1
        col1 += 1
        row2 += 1
        col2 += 1

        left_diff = self.matrix[row2][col1 - 1]
        top_diff = self.matrix[row1-1][col2]
        corner = self.matrix[row1-1][col1-1]

        res = self.matrix[row2][col2]

        return res - left_diff - top_diff + corner


# Your NumMatrix object will be instantiated and called as such:
# obj = NumMatrix(matrix)
# param_1 = obj.sumRegion(row1,col1,row2,col2)